File: cs/StrategyCore/FactorEngine.cs
using System;
using System.Collections.Generic;
using MockTrader.Data;

namespace MockTrader.Core
{
    public sealed class FactorParams
    {
        public int MomentumLookback { get; set; } = 120;
        public int MomentumSkip { get; set; } = 21;
        public int AmihudWindow { get; set; } = 20;
        public int VolumeRatioWindow { get; set; } = 20;
        public int SkewWindow { get; set; } = 20;
        public bool RollYieldAnnualize { get; set; } = true;
        public double WinsorizeK { get; set; } = 2.5;
    }

    public sealed class FactorPanel
    {
        public List<string> Dates { get; set; }
        public List<string> Varieties { get; set; }
        public Dictionary<string, Dictionary<string, double?[]>> Raw { get; set; } // factor -> code -> array
        public Dictionary<string, Dictionary<string, double?[]>> Z { get; set; }
    }

    public static class FactorKeys
    {
        public static readonly string[] All = { "momentum", "liquidity", "volume", "skewness", "rollYield" };
    }

    public sealed class FactorEngine
    {
        public FactorPanel Compute(DataAccess ds, FactorParams p = null)
        {
            p = p ?? new FactorParams();
            var varieties = ds.Codes;
            var dates = ds.Dates;
            var raw = new Dictionary<string, Dictionary<string, double?[]>>();
            foreach (var f in FactorKeys.All) raw[f] = new Dictionary<string, double?[]>();

            foreach (var code in varieties)
            {
                var f = ComputeVariety(ds, code, p);
                foreach (var k in FactorKeys.All) raw[k][code] = f[k];
            }

            var z = new Dictionary<string, Dictionary<string, double?[]>>();
            foreach (var k in FactorKeys.All)
                z[k] = CrossSectionalZ(dates, varieties, raw[k], p.WinsorizeK);

            return new FactorPanel { Dates = dates, Varieties = varieties, Raw = raw, Z = z };
        }

        static double?[] Empty(int n) => new double?[n];

        static Dictionary<string, double?[]> ComputeVariety(DataAccess ds, string code, FactorParams p)
        {
            var s = ds.GetSeries(code);
            var dates = ds.Dates;
            int T = dates.Count;
            var outF = new Dictionary<string, double?[]>();
            foreach (var k in FactorKeys.All) outF[k] = Empty(T);

            var ret = Empty(T);
            for (int t = 1; t < T; t++)
            {
                var a = s.MainAdj[t]; var b = s.MainAdj[t - 1];
                if (a != null && b != null && b > 0) ret[t] = a.Value / b.Value - 1.0;
            }

            for (int t = 0; t < T; t++)
            {
                // momentum
                int iEnd = t - p.MomentumSkip, iStart = t - p.MomentumLookback;
                if (iStart >= 0 && iEnd >= 0 && s.MainAdj[iEnd] != null && s.MainAdj[iStart] != null && s.MainAdj[iStart] > 0)
                    outF["momentum"][t] = s.MainAdj[iEnd].Value / s.MainAdj[iStart].Value - 1.0;

                // amihud + liquidity
                int w = p.AmihudWindow;
                if (t >= w)
                {
                    double acc = 0; int cnt = 0;
                    for (int i = t - w + 1; i <= t; i++)
                    {
                        double? dolVol = (s.MainRaw[i] != null && s.MainVol[i] != null) ? s.MainRaw[i].Value * s.MainVol[i].Value : (double?)null;
                        if (ret[i] != null && dolVol != null && dolVol > 0) { acc += Math.Abs(ret[i].Value) / dolVol.Value; cnt++; }
                    }
                    if (cnt > 0) outF["liquidity"][t] = -(acc / cnt);
                }

                // volume ratio
                int vw = p.VolumeRatioWindow;
                if (t >= vw && s.MainVol[t] != null)
                {
                    double sv = 0; int cnt = 0;
                    for (int i = t - vw; i < t; i++)
                        if (s.MainVol[i] != null) { sv += s.MainVol[i].Value; cnt++; }
                    if (cnt > 0 && sv > 0) outF["volume"][t] = s.MainVol[t].Value / (sv / cnt) - 1.0;
                }

                // skewness
                int sw = p.SkewWindow;
                if (t >= sw)
                {
                    var win = new List<double>();
                    for (int i = t - sw + 1; i <= t; i++) if (ret[i] != null) win.Add(ret[i].Value);
                    if (win.Count >= 10) outF["skewness"][t] = Stats.Skewness(win);
                }

                // roll yield
                if (s.MainRaw[t] != null && s.SubRaw[t] != null && s.SubRaw[t] > 0)
                {
                    string mc = s.MainCode[dates[t]];
                    string sc = s.SubCode[dates[t]];
                    if (mc != null && sc != null)
                    {
                        var dm = DataAccess.ParseContractCode(code, mc).Delivery;
                        var dsc = DataAccess.ParseContractCode(code, sc).Delivery;
                        double dtDays = Math.Abs(Util.DiffDays(dm, dsc));
                        double rawRY = (s.MainRaw[t].Value - s.SubRaw[t].Value) / s.SubRaw[t].Value;
                        outF["rollYield"][t] = p.RollYieldAnnualize && dtDays > 0 ? rawRY * (365.0 / dtDays) : rawRY;
                    }
                }
            }
            return outF;
        }

        static Dictionary<string, double?[]> CrossSectionalZ(List<string> dates, List<string> varieties,
            Dictionary<string, double?[]> rawByCode, double winsorizeK)
        {
            var outZ = new Dictionary<string, double?[]>();
            foreach (var code in varieties) outZ[code] = Empty(dates.Count);
            for (int t = 0; t < dates.Count; t++)
            {
                var vals = new List<double>();
                var idx = new List<string>();
                foreach (var code in varieties)
                {
                    var v = rawByCode[code][t];
                    if (v != null && !double.IsNaN(v.Value) && !double.IsInfinity(v.Value)) { vals.Add(v.Value); idx.Add(code); }
                }
                if (vals.Count < 2) continue;
                var w = Stats.Winsorize(vals, winsorizeK);
                var z = Stats.ZScore(w);
                for (int k = 0; k < idx.Count; k++) outZ[idx[k]][t] = z[k];
            }
            return outZ;
        }
    }
}

File: src/core/factors/factorEngine.js
/**
 * factorEngine.js — 5 因子引擎 (S2: FactorEngine "dll")。纯函数、无未来函数、参数可调。
 *
 * 因子与符号约定（更高 = 更倾向做多；符号可在 S3 配置覆盖）：
 *  1. momentum 截面动量：过去 lookback 日收益，跳过最近 skip 日（默认 120/21，即"12-1 动量"）。
 *  2. liquidity 流动性：-Amihud 非流动性（越高越流动）。Amihud = mean(|ret| / 成交额)。
 *  3. volume 成交量：量比 = 当日成交量 / 过去 ratioWindow 日均量 - 1。
 *  4. skewness 价格偏度：window 日收益率偏度（默认 20）。
 *  5. rollYield 展期收益率：(主力价 - 次主力价)/次主力价，年化（Backwardation 为正）。
 *
 * 所有因子仅用 t 日及之前数据（动量/偏度/流动性用历史窗口；展期收益率用 t 日收盘的
 * 主力与次主力横截面价差，交割日由合约代码决定，非未来信息）。
 * 截面标准化：逐日对各品种因子值 winsorize 后 z-score。
 */

import { mean, std, winsorize, zscore } from '../utils.js';
import { parseContractCode } from '../data/dataAccess.js';

export const DEFAULT_FACTOR_PARAMS = {
  momentum: { lookback: 120, skip: 21 },
  liquidity: { amihudWindow: 20 },
  volume: { ratioWindow: 20, momentumLookback: 60 },
  skewness: { window: 20 },
  rollYield: { annualize: true },
  winsorizeK: 2.5,
  crossSectionalZ: true,
};

/** 因子符号默认值（更高 = 更倾向做多） */
export const FACTOR_SIGNS = { momentum: 1, liquidity: 1, volume: 1, skewness: 1, rollYield: 1 };

/** 样本偏度 */
export function skewness(arr) {
  const n = arr.length;
  if (n < 3) {
    return NaN;
  }
  const m = mean(arr);
  const s = std(arr, 0);
  if (!(s > 0)) {
    return NaN;
  }
  let acc = 0;
  for (let i = 0; i < n; i++) {
    acc += Math.pow(arr[i] - m, 3);
  }
  return acc / n / Math.pow(s, 3);
}

function asArray(n) {
  return new Array(n).fill(null);
}

/**
 * 计算单品种各因子原始序列（与 dates 对齐，无数据处为 null）。
 */
export function computeVarietyFactors(ds, code, params) {
  const series = ds.getSeries(code);
  const dates = ds.dates;
  const T = dates.length;
  const p = params;
  const mom = p.momentum;
  const liq = p.liquidity;
  const vol = p.volume;
  const sk = p.skewness;
  const ry = p.rollYield;

  const out = {
    momentum: asArray(T),
    liquidity: asArray(T),
    volume: asArray(T),
    skewness: asArray(T),
    rollYield: asArray(T),
    turnover: asArray(T),
    amihud: asArray(T),
  };

  // 主连续（后复权）日收益：用于动量/偏度/Amihud
  const ret = asArray(T);
  for (let t = 1; t < T; t++) {
    const a = series.mainAdj[t];
    const b = series.mainAdj[t - 1];
    if (a != null && b != null && b > 0) {
      ret[t] = a / b - 1;
    }
  }

  for (let t = 0; t < T; t++) {
    // ---- momentum (skip 近 1 月) ----
    const iEnd = t - mom.skip;
    const iStart = t - mom.lookback;
    if (
      iStart >= 0 &&
      iEnd >= 0 &&
      series.mainAdj[iEnd] != null &&
      series.mainAdj[iStart] != null &&
      series.mainAdj[iStart] > 0
    ) {
      out.momentum[t] = series.mainAdj[iEnd] / series.mainAdj[iStart] - 1;
    }

    // ---- liquidity: turnover & Amihud ----
    if (
      series.mainRaw[t] != null &&
      series.mainVol[t] != null &&
      series.mainOi[t] != null &&
      series.mainOi[t] > 0
    ) {
      out.turnover[t] = (series.mainRaw[t] * series.mainVol[t]) / series.mainOi[t];
    }
    const w = liq.amihudWindow;
    if (t >= w) {
      let acc = 0;
      let cnt = 0;
      for (let i = t - w + 1; i <= t; i++) {
        const dolVol =
          series.mainRaw[i] != null && series.mainVol[i] != null
            ? series.mainRaw[i] * series.mainVol[i]
            : null;
        if (ret[i] != null && dolVol != null && dolVol > 0) {
          acc += Math.abs(ret[i]) / dolVol;
          cnt++;
        }
      }
      if (cnt > 0) {
        out.amihud[t] = acc / cnt;
      }
    }
    out.liquidity[t] = out.amihud[t] != null ? -out.amihud[t] : null;

    // ---- volume: 量比 ----
    const vw = vol.ratioWindow;
    if (t >= vw && series.mainVol[t] != null) {
      let s = 0;
      let cnt = 0;
      for (let i = t - vw; i < t; i++) {
        if (series.mainVol[i] != null) {
          s += series.mainVol[i];
          cnt++;
        }
      }
      if (cnt > 0 && s > 0) {
        out.volume[t] = series.mainVol[t] / (s / cnt) - 1;
      }
    }

    // ---- skewness ----
    const sw = sk.window;
    if (t >= sw) {
      const win = [];
      for (let i = t - sw + 1; i <= t; i++) {
        if (ret[i] != null) {
          win.push(ret[i]);
        }
      }
      if (win.length >= 10) {
        out.skewness[t] = skewness(win);
      }
    }

    // ---- roll yield ----
    if (series.mainRaw[t] != null && series.subRaw[t] != null && series.subRaw[t] > 0) {
      const mc = series.mainCode[dates[t]];
      const sc = series.subCode[dates[t]];
      if (mc && sc) {
        const dm = parseContractCode(code, mc).delivery;
        const dsc = parseContractCode(code, sc).delivery;
        const dtDays = Math.abs((Date.parse(dsc) - Date.parse(dm)) / 86400000);
        const raw = (series.mainRaw[t] - series.subRaw[t]) / series.subRaw[t];
        out.rollYield[t] = ry.annualize !== false && dtDays > 0 ? raw * (365 / dtDays) : raw;
      }
    }
  }
  return out;
}

/**
 * 截面标准化：对某因子逐日 winsorize + z-score。
 * @returns {Object<string, Array<number|null>>} code -> z 数组
 */
export function crossSectionalZ(dates, varieties, rawByCode, winsorizeK) {
  const out = {};
  for (const code of varieties) {
    out[code] = asArray(dates.length);
  }
  for (let t = 0; t < dates.length; t++) {
    const vals = [];
    const idx = [];
    for (const code of varieties) {
      const v = rawByCode[code][t];
      if (v != null && Number.isFinite(v)) {
        vals.push(v);
        idx.push(code);
      }
    }
    if (vals.length < 2) {
      continue;
    }
    const w = winsorize(vals, winsorizeK);
    const z = zscore(w);
    for (let k = 0; k < idx.length; k++) {
      out[idx[k]][t] = z[k];
    }
  }
  return out;
}

export class FactorEngine {
  /**
   * 计算全品种因子面板。
   * @returns {{dates, varieties, raw, z, aux, params, signs}}
   */
  compute(ds, params = {}) {
    const p = Object.assign({}, DEFAULT_FACTOR_PARAMS, params, {
      momentum: Object.assign({}, DEFAULT_FACTOR_PARAMS.momentum, params.momentum),
      liquidity: Object.assign({}, DEFAULT_FACTOR_PARAMS.liquidity, params.liquidity),
      volume: Object.assign({}, DEFAULT_FACTOR_PARAMS.volume, params.volume),
      skewness: Object.assign({}, DEFAULT_FACTOR_PARAMS.skewness, params.skewness),
      rollYield: Object.assign({}, DEFAULT_FACTOR_PARAMS.rollYield, params.rollYield),
    });
    const varieties = ds.codes;
    const dates = ds.dates;
    const raw = { momentum: {}, liquidity: {}, volume: {}, skewness: {}, rollYield: {} };
    const aux = { turnover: {}, amihud: {} };
    for (const code of varieties) {
      const f = computeVarietyFactors(ds, code, p);
      for (const k of ['momentum', 'liquidity', 'volume', 'skewness', 'rollYield']) {
        raw[k][code] = f[k];
      }
      aux.turnover[code] = f.turnover;
      aux.amihud[code] = f.amihud;
    }
    const z = {};
    for (const k of ['momentum', 'liquidity', 'volume', 'skewness', 'rollYield']) {
      z[k] = crossSectionalZ(dates, varieties, raw[k], p.winsorizeK);
    }
    return { dates, varieties, raw, z, aux, params: p, signs: Object.assign({}, FACTOR_SIGNS) };
  }
}
